pkg/resources/management.cattle.io/v3/roletemplate/mutator.go

```diff
@@ -0,0 +1,67 @@
+package roletemplate
+
+import (
+	"fmt"
+
+	"github.com/rancher/webhook/pkg/admission"
+	v3 "github.com/rancher/webhook/pkg/generated/controllers/management.cattle.io/v3"
+	objectsv3 "github.com/rancher/webhook/pkg/generated/objects/management.cattle.io/v3"
+	"github.com/rancher/webhook/pkg/patch"
+	admissionv1 "k8s.io/api/admission/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/trace"
+)
+
+const cleanupKey = "cleanup.cattle.io/rtUpgradeCluster"
+
+// Mutator implements admission.MutatingAdmissionWebhook.
+type Mutator struct {
+	globalRoles v3.GlobalRoleCache
+}
+
+// NewMutator returns a new mutator for GlobalRoleBindings.
+func NewMutator(grCache v3.GlobalRoleCache) *Mutator {
+	return &Mutator{
+		globalRoles: grCache,
+	}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (m *Mutator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this mutator.
+func (m *Mutator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{admissionregistrationv1.Create}
+}
+
+// MutatingWebhook returns the MutatingWebhook used for this CRD.
+func (m *Mutator) MutatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.MutatingWebhook {
+	mutatingWebhook := admission.NewDefaultMutatingWebhook(m, clientConfig, admissionregistrationv1.ClusterScope, m.Operations())
+	return []admissionregistrationv1.MutatingWebhook{*mutatingWebhook}
+}
+
+// Admit handles the webhook admission request sent to this webhook.
+// If this function is called without NewMutator(..) calls will panic.
+func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	listTrace := trace.New("RoleTemplate Mutator Admit", trace.Field{Key: "user", Value: request.UserInfo.Username})
+	defer listTrace.LogIfLong(admission.SlowTraceDuration)
+
+	newGRB, err := objectsv3.RoleTemplateFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get RoleTemplate from request: %w", err)
+	}
+	if newGRB.Annotations == nil {
+		newGRB.Annotations = map[string]string{}
+	}
+	newGRB.Annotations[cleanupKey] = "true"
```

Comment:
Could you explain why this can't be done by a controller afterwards? Something like an annotation seems ideal for the controller.